### PR TITLE
Add Foldable1 And Traversable1 type classes

### PR DESCRIPTION
**Pull request by LukaJCB**

### CODE REVIEW

File: src/Data/Semigroup/Foldable.purs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+-- |
+-- | Default implementations are provided by the following functions:
+-- |
+-- | - `fold1Default`
+-- | - `foldMap1Default`
+-- |
+-- | Note: some combinations of the default implementations are unsafe to
+-- | use together - causing a non-terminating mutually recursive cycle.
+-- | These combinations are documented per function.
+class (Foldable t) <= Foldable1 t where
```
Comment (paf31) [inline src/Data/Semigroup/Foldable.purs]: Nitpick: parens are redundant here, and on the lines below.

File: src/Data/Semigroup/Foldable.purs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+instance foldableDual :: Foldable1 Dual where
+  foldMap1 f (Dual x) = f x
+  fold1 = fold1Default
+
+instance foldableMultiplicative :: Foldable1 Multiplicative where
+  foldMap1 f (Multiplicative x) = f x
+  fold1 = fold1Default
+
+newtype Act f a = Act (f a)
```
Comment (paf31) [inline src/Data/Semigroup/Foldable.purs]: Does this need to be exported?

Comment (LukaJCB) [inline src/Data/Semigroup/Foldable.purs]: Great catch, it does not :)

File: src/Data/Semigroup/Foldable.purs (hunk 1 above)
Comment (LukaJCB) [inline src/Data/Semigroup/Foldable.purs]: Done!

File: src/Data/Semigroup/Foldable.purs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+-- | Default implementations are provided by the following functions:
+-- |
+-- | - `fold1Default`
+-- | - `foldMap1Default`
+-- |
+-- | Note: some combinations of the default implementations are unsafe to
+-- | use together - causing a non-terminating mutually recursive cycle.
+-- | These combinations are documented per function.
+class Foldable t <= Foldable1 t where
+  foldMap1 :: forall a m. (Semigroup m) => (a -> m) -> t a -> m
```
Comment (paf31) [inline src/Data/Semigroup/Foldable.purs]: redundant parens here too (to the left of `=>`)

File: src/Data/Semigroup/Foldable.purs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+instance foldableMultiplicative :: Foldable1 Multiplicative where
+  foldMap1 f (Multiplicative x) = f x
+  fold1 = fold1Default
+
+newtype Act f a = Act (f a)
+
+getAct :: forall f a. Act f a -> f a
+getAct (Act f) = f
+
+instance semigroupAct :: (Apply f) => Semigroup (Act f a) where
```
Comment (paf31) [inline src/Data/Semigroup/Foldable.purs]: one more

**Comment by paf31** [comment]
👍 LGTM. Any comments from anyone else before merging? @garyb @hdgarrood @LiamGoodacre

**Comment by LiamGoodacre** [comment]
Yeah I'm not too hot on naming things with 1, but I've never thought of anything better.

**Comment by LukaJCB** [comment]
Honestly the name comes from the Haskell equivalent, so I can't claim to have come up with it. An alternative might be `NonEmptyFoldable` or something of the like, but it's not perfect either.

**Comment by garyb** [comment]
> Yeah I'm not too hot on naming things with 1, but I've never thought of anything better.

There's always priming (no, don't do that really) 😛 

> An alternative might be `NonEmptyFoldable` or something of the like, but it's not perfect either.

I did think about suggesting that, but I think `1` is fine here, since the 1-suffixed functions totally make sense, and a quick look at the types reveals exactly what's going on anyway.

**Comment by LiamGoodacre** [comment]
`Semigroupable` 😆

**Comment by paf31** [comment]
Thanks!

**Pull request merged**